### ISSUE TITLE
Optional dependencies

### ISSUE THREAD

**Issue by jakirkham** [conda/conda#3299]
This is a mixture of a `conda` and `conda-build` issue. However, I'm going to start by raising it here. We can latter break it up into smaller issues on the relevant repos. For now, I simply want to get the discussion going.

Periodically one has optional dependencies. The initial thought on this was that one could try to build against all dependencies and ship only the required subset (if that would work). Sometimes this works if a package has a GUI add-on requiring Qt. Due to the need for version pinning, this is often unsatisfactory as there is no way to constrain the version of Qt that does get installed later. However, sometimes it doesn't work without considering version pinning (e.g. building `libxml2` support for something).

While one could consider the use of features to help here, generally the view has become that features are extremely unsavory. This involves how they are handled in the solver. Though it also adds another package to handle a problem that may only affect one package in the first place. It would be much better if there was some way we could specify these optional features in the recipe.

Please share your thoughts

cc @msarahan @pelson @ocefpaf

**Comment by kalefranz** [comment]
>  Due to the need for version pinning

This is the key issue of why optional dependencies are important.

Note this isn't exactly the same thing as optional dependencies in the pip/setuptools sense with the square bracket syntax.  I haven't yet seen a compelling use case for adding optional dependencies in the "square bracket" sense.

However, this issue that @jakirkham raises is to me clearly legitimate.

**Comment by pelson** [comment]
xref: #2984

>  I haven't yet seen a compelling use case for adding optional dependencies

The following work for me: ```conda install matplotlib[qt]``` and ```conda install jupyter[qtconsole]```
I could leave the square bracket syntax, but the capability to be able to express "now install the matplotlib qt/tk/gtk/notebook backend" independently is the end-point that I would want to expose to a user.

**Comment by jakirkham** [comment]
So the first case feels exactly like what I'm trying to point out here.

The second case not so much as it seems like that could easily resolve to `conda install qtconsole` unless I'm missing something crucial.

> the capability to be able to express "now install the matplotlib qt/tk/gtk/notebook backend" independently is the end-point that I would want to expose to a user.

Exactly. As this will require pinnings being matched, this is one such case I'm thinking of.

**Comment by Korijn** [comment]
Another example is VTK which has to be built against the desired backend, e.g. Qt or osmesa. The backend then is not an optional dependency, but rather a flavor of VTK that you are installing.

**Comment by jakirkham** [comment]
A different example that is Python only, but involves a post-link script, would be `ipyparallel` where we want to enable a widget for the Jupyter Notebook, but only if `notebook` is being installed or present in the directory. As we can't do that, we are left to check for the `notebook` in a kind of hacky way ( https://github.com/conda-forge/ipyparallel-feedstock/pull/6 ). Admittedly approaching this use case opens us up to considering [un]install order a bit more closely with optional dependencies.

cc @minrk

**Comment by kalefranz** [comment]
This enhancement is definitely on the conda roadmap. There are many high priorities at the moment though.

**Comment by jakirkham** [comment]
Any chance we might get to see that roadmap? 😉 I expect that will be very insightful for both sides of the table.

**Comment by kalefranz** [comment]
Yes. It's on my list of things to do. I'll clean it up and put it on a wiki page or something like that.

**Comment by jakirkham** [comment]
IMHO a repo might be better, but just having it visible will already be a good thing.

**Comment by astrofrog** [comment]
I have a use case for this too - I am the lead developer of the glueviz package, which be used in many scientific domains. For each domain there is a set of dependencies that can be installed (for example ``pydicom`` for medical scientists). At the moment, I am just adding all optional dependencies to the conda recipe, but over time, this will mean that the average medical scientist that wants to install glue will also get many unrelated packages/plugins. I'd therefore like a way for users to be able to do:

```
conda install glueviz[astronomy]
conda install glueviz[medical]
```

Creating meta-packages for each of these is also going to be unmanageable because over time there might be say 10 different fields of research concerned. So I would like a way to define these different groups of dependencies straight in the recipe.

**Comment by jakirkham** [comment]
Another interesting use case is packages that have dependencies where they wish to include some option dependencies. For instance, `lektor` requires `requests[security]`, but we don't have a good way to specify this yet. So we end up including `requests` and its optional dependencies in `lektor`.

xref: https://github.com/lektor/lektor/blob/2.3/setup.py#L27
xref: https://github.com/conda-forge/staged-recipes/pull/1836

cc @goanpeca

**Comment by kalefranz** [comment]
I want to disambiguate between two different types of "optional dependencies."  

The type most here might be familiar with are the pip type, denoted by square brackets.  Let's call those "aliased group dependencies."  The single alias refers to a group of additional dependencies.

The second type I'd like to call "conditional dependencies."  This type in fact is completely different and orthogonal to the first.  This type solves the problem of why the conda-forge `django` package has a dependency on `gdal`.  This type tells the conda solver, for example, *if `gdal` is going to be installed in this environment it must conform to a certain version specification.  The conda-forge `django` package would thus ideally declare this conditional dependency in it's meta.yaml.

Two separate types of optional dependencies.  I fully support the latter, and it's on the conda roadmap.  (I know...  I just need to clean it up so I can make it public.)  The first type I'm not completely onboard with yet.  With the second type properly implemented, users are free to add whatever additional dependencies they want, while the package maintainer is still guaranteed the ability to control the versions of *optional* dependencies that end up in the environment.

**Comment by Korijn** [comment]
It's been four months now, any news on the roadmap?

Also, I think the VTK scenario I shared earlier in this thread is not covered by your latest comment, @kalefranz

**Comment by jakirkham** [comment]
Agreed, any news on this @kalefranz?

**Comment by astrofrog** [comment]
I do feel that what @kalefranz calls 'aliased group dependencies' would be important to support - at the moment it's essentially impossible to replicate the installation options for a package that uses ``extras_require`` in conda. Either one sticks to ``install_requires`` (as is done in the ``defaults`` channel), or one can adopt a reasonably flexible policy (conda-forge style) of allowing optional dependencies to be declared as full dependencies. I feel strongly that the square bracket notation that works for pip should work seamlessly for conda as well, since this will be the most intuitive to users.

**Comment by kalefranz** [comment]
Constrained optional dependencies are in conda 4.4 as #4982.

**Comment by kalefranz** [comment]
Circling back on this topic for the concept of aliases for sets of packages, we actually already have a long-established facility to do that with conda.  And it's used all the time.  The answer is the conda metapackage.  No, it's not exactly like pip does it.  But the functionality is equivalent.

**Comment by kalefranz** [comment]
@jbednar Why don't you open a new feature request issue, replicating the bulk of your comments, and referencing this issue.  This issue was closed with 4.4 constrained dependencies, but I understand it wasn't a solution for the use case you describe.  Best to track the unresolved request in an open issue rather than a closed one.

**Comment by jbednar** [comment]
Ok, I moved the discussion from the above into #7502.  Note that the request for documentation of the previously implemented approach still stands.
